pkg/internal/source/informer.go

```diff
@@ -0,0 +1,176 @@
+	if is.GetInformer == nil {
+		return fmt.Errorf("must specify Informer.GetInformer")
+	}
+
+	is.mu.Lock()
+	defer is.mu.Unlock()
+	if is.started != nil {
+		return fmt.Errorf("Informer source already started")
+	}
+
+	if is.canceled {
+		return nil
+	}
+
+	is.started = make(chan error, 1)
+	var startCtx, eventHandlerCtx context.Context
+	startCtx, is.startCancel = context.WithCancel(ctx)
+	eventHandlerCtx, is.eventHandlerCancel = context.WithCancel(context.Background())
+	go func() {
+		defer close(is.started)
+
+		// Get the informer.
+		informer, err := is.GetInformer(startCtx)
+		if err != nil {
+			is.started <- err
+			return
+		}
+
+		// Store the informer and add the event handler.
+		func() {
+			is.mu.Lock()
+			defer is.mu.Unlock()
+
+			is.informer = informer
+
+			if is.WaitForInformerSync == nil {
+				is.WaitForInformerSync = func(informerSyncCtx context.Context) error {
+					return wait.PollImmediateUntil(
+						syncedPollPeriod,
+						func() (bool, error) { return is.informer.HasSynced(), nil },
+						informerSyncCtx.Done(),
+					)
+				}
+			}
+
+			eventHandlerRegistration, err := is.informer.AddEventHandler(NewEventHandler(eventHandlerCtx, queue, handler, prct))
+			if err != nil {
+				is.started <- err
+				return
+			}
+			is.eventHandlerRegistration = eventHandlerRegistration
+		}()
+
+		if err := is.WaitForInformerSync(startCtx); err != nil {
+			// Remove the event handler.
+			is.mu.Lock()
+			if is.eventHandlerRegistration != nil {
+				is.eventHandlerCancel()
+				err = kerrors.NewAggregate([]error{err, is.informer.RemoveEventHandler(is.eventHandlerRegistration)})
+			}
+			is.mu.Unlock()
+
+			is.started <- err
+			return
+		}
+
+		// Informer is ready.
+		is.started <- nil
+	}()
+
+	return nil
+}
+
+func (is *Informer) String() string {
+	return fmt.Sprintf("informer source: %p", is.GetInformer)
+}
+
+// WaitForSync implements SyncingSource to allow controllers to wait with starting
+// workers until the cache is synced.
+func (is *Informer) WaitForSync(ctx context.Context) error {
+	select {
+	case err := <-is.started:
+		return err
+	case <-ctx.Done():
+		is.startCancel()
+		<-is.started // wait for the goroutine to finish; TODO: do something with the error?
+		if errors.Is(ctx.Err(), context.Canceled) {
+			return nil
+		}
+
+		return fmt.Errorf("timed out waiting for cache to be synced")
+	}
+}
+
+// Stop implements StoppableSource to stop it dynamically.
+func (is *Informer) Stop() error {
+	is.mu.Lock()
+	defer is.mu.Unlock()
+
+	if is.canceled {
+		return nil
+	}
+	is.canceled = true
+
+	// Return if it has not been started.
+	if is.started == nil {
+		return nil
+	}
+
+	select {
+	case <-is.started:
+	default:
+		is.startCancel()
+
+		is.mu.Unlock()
+		// Wait for starting abort
+		<-is.started
+		is.mu.Lock()
+	}
+
+	if is.eventHandlerRegistration != nil {
+		is.eventHandlerCancel()
+		return is.informer.RemoveEventHandler(is.eventHandlerRegistration)
```

Comment:
To re-raise a thread from another pr, it looks like this doesn't tear down the informer if there are no longer any event handlers:

https://github.com/kubernetes-sigs/controller-runtime/pull/2099#discussion_r1051660956

IIRC the comment on the old PR was cache cleanup would be a follow-on change. Is that the same idea here?

Reply:
Yes, I think the informer has to be responsible for dynamically starting and shutting down when there are no EventHandlers registered anymore.

Reply:
I'm okay with dynamic, but having a manual method as an MVP may be a lower bar.